### PR TITLE
Add exclude private URLs feature

### LINKED ISSUE #5: Exclude private URLs
We should have flags for excluding private, link-local, and loopback hosts.

Basically this means creating the new commandline flags and mapping them to the functions in std::net (linked below):

* `--exclude-private`: [Exclude private addresses](https://doc.rust-lang.org/std/net/struct.Ipv4Addr.html#method.is_private)
* `--exclude-link-local`: [Exclude link-local addresses](https://doc.rust-lang.org/std/net/struct.Ipv4Addr.html#method.is_link_local)
* `--exclude-loopback`: [Exclude loopback](https://doc.rust-lang.org/std/net/struct.Ipv4Addr.html#method.is_loopback)

Newcomers welcome!

### PR DESCRIPTION
**Pull request by pawroman**
Currently, the excludes are based purely on IP address ranges and only work for websites (no support for mail addresses). The IPv6 support is implemented, to a degree possible with Rust standard library as of 1.47.0. See commit messages and code comments for more details.

Note: while IP address exclude works, it's possible that some URLs will resolve to link-local, private or loopback addresses. While this is not very likely, it's a possibility. To address this, we'd have to DNS-resolve all URIs and then run them through IP exclude rules. I have skipped this part because of extra effort involved.

Closes #5.

### CODE REVIEW

**Comment by pawroman** [comment]
I quite honestly don't know why the build is failing. The tests run on my local machine just fine (Rust 1.47.0, Linux, x86_64).

**Comment by pawroman** [comment]
The CLI options parser generated weird flag names for this:

```
Usage: target/debug/lychee [OPTIONS]

Positional arguments:
  inputs                 Input files

Optional arguments:
  -h, --help             show help
  -v, --verbose          Verbose program output
  -m, --max-redirects MAX-REDIRECTS
                         Maximum number of allowed redirects (default: 10)
  -t, --threads THREADS  Number of threads to utilize (defaults to  number of cores available to the system
  -u, --user-agent USER-AGENT
                         User agent (default: curl/7.71.1)
  -i, --insecure         Proceed for server connections considered insecure (invalid TLS) (default: false)
  -s, --scheme SCHEME    Only test links with given scheme (e.g. https)
  -e, --exclude EXCLUDE  Exclude URLs from checking (supports regex)
  -E, --exclude-private  Exclude private IP address ranges from checking
  --exclude-link-local   Exclude link-local IP address range from checking
  --exclude-loopback     Exclude loopback IP address range from checking
  -H, --headers HEADERS  Custom request headers
  -a, --accept ACCEPT    Comma-separated list of accepted status codes for valid links
  -T, --timeout TIMEOUT  Website timeout from connect to response finished (default: 20)
  -M, --method METHOD    Request method (default: get)
```
Any preferences as what to should the flags be?

I have a proposal:
- Drop the short flags, leaving only long ones AND
- Have a `--exclude-all-private` flag, mapped to `-E` that would enable all of them (`--exclude-private --exclude-link-local --exclude-loopback`).

What do you think?

**Comment by pawroman** [comment]
> I quite honestly don't know why the build is failing. The tests run on my local machine just fine (Rust 1.47.0, Linux, x86_64).

I was able to work out that it uses Rust 1.46.0, as included in `ubuntu-latest` image: https://github.com/actions/virtual-environments -- see: https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu1804-README.md 

I have installed Rust 1.46.0 locally using rustup, and re-ran the tests, and they did fail indeed, with the same message.
Then I checked out the `master` branch (without my changes) and re-ran the tests, they failed.

Then I installed Rust 1.45.2 and re-ran the tests for `master` branch. It did not fail. Then I checked my branch, it did not fail either.

So, it's likely a problem with Rust 1.46.0... Some compiler bug perhaps? I'm not sure what the next steps should be. Let me know what you think!

PS: For reference, here's how to install toolchains:

```
$ rustup toolchain install 1.45.2
$ rustup default 1.45.2
```

**Comment by pawroman** [comment]
For the record: adding

```rust
#![type_length_limit = "7912782"]
```

At the very top of `main.rs` makes it work with Rust 1.46.0.

**Comment by mre** [comment]
@xiaochuanyu found a fix in #6. As you guessed, it's likely an issue with the 1.46.0 toolchain. Updating to the latest version seems to fix the issue.

File: src/checker.rs [hunk 1]
```diff
@@ -191,17 +223,44 @@ impl Checker {
...
         false
     }
 
+    fn in_ip_excludes(&self, uri: &Uri) -> bool {
+        if let Some(ipaddr) = uri.host_ip() {
+            if self.excludes.loopback_ips && ipaddr.is_loopback() {
+                return true;
+            }
+
+            // Note: in a pathological case, an IPv6 address can be IPv4-mapped
```
Comment (mre) [inline src/checker.rs:241]: Oh wow. That's both shocking and hilarious. 😆

File: src/checker.rs [hunk 2]
```diff
@@ -285,12 +363,14 @@ mod test {
         checker
     }
 
+    fn website_url(s: &str) -> Uri {
```
Comment (mre) [inline src/checker.rs:366]: nice!

File: src/extract.rs [hunk 3]
```diff
@@ -23,6 +24,17 @@ impl Uri {
             Uri::Mail(_address) => None,
         }
     }
+
+    pub fn host_ip(&self) -> Option<IpAddr> {
```
Comment (mre) [inline src/extract.rs:28]: nicely done

**Comment by mre** [comment]
@pawroman can you reintegrate `master`? Merged the changes by @xiaochuanyu. 
Apart from that it looks great!

**Comment by pawroman** [comment]
> @pawroman can you reintegrate `master`? Merged the changes by @xiaochuanyu.
> Apart from that it looks great!

Thanks for the review. I have rebased against master, good to see that the CI is green now :)

Any thoughts on the CLI options? https://github.com/hello-rust/lychee/pull/7#issuecomment-706229204

**Comment by mre** [comment]
> I have a proposal:
> * Drop the short flags, leaving only long ones AND
> * Have a --exclude-all-private flag, mapped to -E, that would enable all of them (--exclude-private --exclude-link-local --exclude-loopback).

Sounds very reasonable to me. 👍

**Comment by pawroman** [comment]
> Sounds very reasonable to me. +1

:+1: 

Implemented, along with a test that runs the program against a test file with the flag.

This should now be ready to merge.

**Comment by mre** [comment]
Just gave it a spin locally and it does what it says on the can. Great work and thanks for your contribution @pawroman. 😃

**Pull request merged**